File: Runtime/OutputStratumTransitionCollection.vb
'*********************************************************************************************
' ST-Sim: A SyncroSim Module for the ST-Sim State-and-Transition Model.
'
' Copyright © 2007-2017 Apex Resource Management Solution Ltd. (ApexRMS). All rights reserved.
'
'*********************************************************************************************

Imports SyncroSim.Common
Imports System.Collections.ObjectModel

Friend Class OutputStratumTransitionCollection
    Inherits KeyedCollection(Of SevenIntegerLookupKey, OutputStratumTransition)

    Public Sub New()
        MyBase.New(New SevenIntegerLookupKeyEqualityComparer)
    End Sub

    Protected Overrides Function GetKeyForItem(ByVal item As OutputStratumTransition) As SevenIntegerLookupKey

        Return New SevenIntegerLookupKey(
            item.StratumId,
            LookupKeyUtils.GetOutputCollectionKey(item.SecondaryStratumId),
            LookupKeyUtils.GetOutputCollectionKey(item.TertiaryStratumId),
            item.Iteration,
            item.Timestep,
            item.TransitionGroupId,
            item.AgeKey)

    End Function

End Class

File: src/Runtime/OutputStratumTransitionCollection.cs
// stsim: A SyncroSim Package for developing state-and-transition simulation models using ST-Sim.
// Copyright © 2007-2021 Apex Resource Management Solutions Ltd. (ApexRMS). All rights reserved.

using SyncroSim.Common;
using System.Collections.ObjectModel;

namespace SyncroSim.STSim
{
    internal class OutputStratumTransitionCollection : KeyedCollection<EightIntegerLookupKey, OutputStratumTransition>
    {
        public OutputStratumTransitionCollection() : base(new EightIntegerLookupKeyEqualityComparer())
        {
        }

        protected override EightIntegerLookupKey GetKeyForItem(OutputStratumTransition item)
        {
            return new EightIntegerLookupKey(
                item.StratumId, 
                LookupKeyUtils.GetOutputCollectionKey(item.SecondaryStratumId), 
                LookupKeyUtils.GetOutputCollectionKey(item.TertiaryStratumId), 
                item.Iteration, 
                item.Timestep, 
                item.TransitionGroupId, 
                item.AgeKey, 
                item.EventIdKey);
        }
    }
}
